### PR TITLE
[update-price-p2] add cumulative func tree

### PR DESCRIPTION
**Pull request by ninjaahhh**
next part: 
1. remove cumulative liquidity when unstake
2. reward calculation based on `_cumulativeAccumulatedRewards`

### CODE REVIEW

File: contracts/UniversalV3Staker.sol [hunk 1]
```diff
@@ -355,9 +371,21 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
         lastTick = tick;
 
-        // TODO: calculate eligible rewards based on liquidity
+        uint208 liquidityLower = _cumulativeLiquidityLower.get(cfNbits, tickBeforeUpdate);
+        uint208 liquidityUpper = _cumulativeLiquidityUpper.get(cfNbits, tickBeforeUpdate);
+        uint208 liquidity = liquidityLower - liquidityUpper;
+        require(liquidity <= liquidityLower, 'UniswapV3Staker::updatePrice: overflow');
+        if (liquidity == 0) return;
+
+        uint256 rewardShare = calculatedRewards / uint256(liquidity);
```
Comment (qizhou) [inline contracts/UniversalV3Staker.sol]: We may need to times uint128_max so that rewardShare is not underflowed?

Comment (ninjaahhh) [reply contracts/UniversalV3Staker.sol]: if so, will we have problems when casting it to `uint208`?

Comment (qizhou) [reply contracts/UniversalV3Staker.sol]: Perhaps times uint64_max is enough - this means that the reward function can be 208 - 64 bits, which has 2 ** 144 space.

File: contracts/UniversalV3Staker.sol [hunk 2]
```diff
@@ -355,9 +371,21 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
+        uint208 liquidityLower = _cumulativeLiquidityLower.get(cfNbits, tickBeforeUpdate);
+        uint208 liquidityUpper = _cumulativeLiquidityUpper.get(cfNbits, tickBeforeUpdate);
+        uint208 liquidity = liquidityLower - liquidityUpper;
+        require(liquidity <= liquidityLower, 'UniswapV3Staker::updatePrice: overflow');
+        if (liquidity == 0) return;
+
+        // avoid underflow
+        uint256 rewardShareX64 = (calculatedRewards << 64) / uint256(liquidity);
+        require(uint256(uint208(rewardShareX64)) == rewardShareX64, 'UniswapV3Staker::updatePrice: casting');
+        // i.e. using 207 - 64 = 143 bits to store share
```
Comment (ninjaahhh) [inline contracts/UniversalV3Staker.sol]:
```suggestion
        // i.e. using  208 - 64 = 144 bits to store shares, with the max to be  2 ^ 144 - 1 = ~10^43, thus 10^25 ether
```

File: contracts/UniversalV3Staker.sol [hunk 3]
```diff
@@ -94,16 +107,19 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
     /// @param _maxIncentiveStartLeadTime the max duration of an incentive in seconds
     /// @param _maxIncentiveDuration the max amount of seconds into the future the incentive startTime can be set
+    /// @param _cfNbits cumulative function tree parameter
     constructor(
         IUniswapV3Factory _factory,
         INonfungiblePositionManager _nonfungiblePositionManager,
         uint256 _maxIncentiveStartLeadTime,
-        uint256 _maxIncentiveDuration
+        uint256 _maxIncentiveDuration,
+        uint256 _cfNbits
```
Comment (qizhou) [inline contracts/UniversalV3Staker.sol:116]: I think cfNbits can be determined by predefined tickspace and TICK_MIN/TICK_MAX.

File: contracts/UniversalV3Staker.sol (hunk 1 above)
Comment (ninjaahhh) [reply contracts/UniversalV3Staker.sol]: done

**Review by qizhou — APPROVED** [review]

File: contracts/UniversalV3Staker.sol (hunk 3 above)
Comment (ninjaahhh) [reply contracts/UniversalV3Staker.sol:116]: will do in the next PR, using mapping from different pools to bits

**Pull request merged**